<a href="https://colab.research.google.com/github/CodeWithSridhar/Innomatics-Internship/blob/main/Task5_RegEx.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Task - 5 (RegEx)**

**Questions Link**: https://www.hackerrank.com/domains/python?filters%5Bsubdomains%5D%5B%5D=py-regex 

## Detect Floating Point Number

In [ ]:
import re
class Main():
    def __init__(self):
        self.n = int(input())
        for i in range(self.n):
            self.s = input()
            print(bool(re.match(r'^[-+]?[0-9]*\.[0-9]+$', self.s)))
                    
if __name__ == '__main__':
    Answer = Main()

4
4.0O0
False
-1.00
True
+4.54
True
SomeRandomStuff
False


Explanation:

4.0O0 : O is not a digit.


-1.00 : is valid.

+4.54 : is valid.

SomeRandomStuff: is not a number.

## Group(), Groups() & Groupdict()

In [ ]:
import re

expression = r"([a-zA-Z0-9])\1+"
String = re.search(expression,input('Enter string : '))
if String:
    print(String.group(1))
else:
    print(-1)

Enter string : ..12345678910111213141516171820212223
1


Here, .. is the first repeating character, but it is not alphanumeric.
1 is the first (from left to right) alphanumeric repeating character of the string in the substring 111.

## Re.findall() & Re.finditer()

In [ ]:
import re
String = re.findall(r'(?<=[qwrtypsdfghjklzxcvbnm])([aeiou]{2,})(?=[qwrtypsdfghjklzxcvbnm])', input().strip(), re.IGNORECASE)

if String:
    for i in String:
        print(i)
else:
    print(-1)

rabcdeefgyYhFjkIoomnpOeorteeeeet
ee
Ioo
Oeo
eeeee


Explanation

ee is located between consonant d and f.

Ioo is located between consonant k and m.

Oeo is located between consonant p and r.

eeeee is located between consonant t and t.

## Re.start() & Re.end()

In [1]:
import re

String, k = input(), input()
matches = re.finditer(r'(?=(' + k + '))', String)

Foundanymatch = False
for match in matches:
    Foundanymatch = True
    print((match.start(1), match.end(1) - 1))

if Foundanymatch == False:
    print((-1, -1))

aaadaa
aa
(0, 1)
(1, 2)
(4, 5)


## Regex Substitution

In [ ]:
import re

def Word(any_match):
    if any_match.group(1) == '&&':
        return 'and'
    else:
        return 'or'

for _ in range(int(input('Enter number of lines: '))):
    print(re.sub(r"(?<= )(\|\||&&)(?= )", Word,input('Enter string: ')))

## Validating phone numbers

In [4]:
import re
N = int(input('Enter num of lines: '))

for i in range(N):
    Phone_number = input('Enter your phone number : ')
    if(len(Phone_number)==10 and Phone_number.isdigit()):
        Valid = re.findall(r"^[789]\d{9}$",Phone_number)
        if(len(Valid)==1):
            print("YES")
        else:
            print("NO")
    else:
        print("NO")

Enter num of lines: 2
Enter your phone number : 9587456281
YES
Enter your phone number : 1252478965
NO
